File: ios/rninternals/BiometricAuthModule.m
//
//  BiometricAuthModule.m
//  
//
//  Created by Hazo Baykulov on 07.07.2026.
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_REMAP_MODULE(BiometricAuth, BiometricAuthModule, NSObject)

RCT_EXTERN_METHOD(getBiometryType)
RCT_EXTERN_METHOD(authenticate:(NSString *)reason
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

@end

File: ios/rninternals/BiometricAuthModule.swift
//
//  BiometricAuthModule.swift
//  rninternals
//
//  Created by Hazo Baykulov on 08.07.2026.
//

import LocalAuthentication

@objc(BiometricAuthModule)
class BiometricAuthModule: NSObject {

  @objc
  func getBiometryType() -> String {
    let context = LAContext()
    var error: NSError?
    guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
      return "none"
    }
    switch context.biometryType {
    case .faceID: return "faceID"
    case .touchID: return "touchID"
    default: return "none"
    }
  }

  @objc
  func authenticate(_ reason: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    let context = LAContext()
    var error: NSError?

    guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
      reject("BIOMETRY_UNAVAILABLE", error?.localizedDescription ?? "Biometry not available", error)
      return
    }

     context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason) { success, evalError in
      DispatchQueue.main.async {
        if success {
          resolve(true)
        } else {
          reject("AUTH_FAILED", evalError?.localizedDescription ?? "Authentication failed", evalError)
        }
      }
    }
  }
}
